apps/src/netsim/NetSimVisualization.js

```diff
@@ -0,0 +1,475 @@
+  // Begin by marking all entities background (unvisited)
+  this.entities_.forEach(function (vizEntity) {
+    vizEntity.speculativeIsForeground = false;
+  });
+
+  // Use a simple stack for our list of nodes that need visiting.
+  // If we have a local node, push it onto the stack as our starting point.
+  // (If we don't have a local node, the next step is REALLY EASY)
+  var exploreStack = [];
+  if (this.localNode) {
+    exploreStack.push(this.localNode);
+  }
+
+  // While there are still nodes that need visiting,
+  // visit the next node, marking it as "foreground/visited" and
+  // pushing all of its unvisited connections onto the stack.
+  var currentVizEntity;
+  while (exploreStack.length > 0) {
+    currentVizEntity = exploreStack.pop();
+    this.visitEntityToSetForeground_(currentVizEntity, exploreStack);
+  }
+
+  // Now, we have to correct "depth" for all nodes.
+  // Move all nodes to their new, correct layers
+  var foreground = this.svgRoot_.find('#foreground_group');
+  var background = this.svgRoot_.find('#background_group');
+  var newParent, isForeground;
+  this.entities_.forEach(function (vizEntity) {
+    newParent = undefined;
+    isForeground = $.contains(foreground[0], vizEntity.getRoot()[0]);
+
+    // Check whether a change should occur.  If not, we leave
+    // newParent undefined so that we don't make unneeded DOM changes.
+    if (vizEntity.speculativeIsForeground && !isForeground) {
+      newParent = foreground;
+    } else if (!vizEntity.speculativeIsForeground && isForeground) {
+      newParent = background;
+    }
+
+    // If we do need a DOM change, detach the entity and reattach it to the
+    // new layer.
+    // Special rule (for now): Prepend wires so that they show up behind
+    // nodes.  Will need a better solution for this if/when the viz gets
+    // more complex.
+    if (newParent) {
+      vizEntity.getRoot().detach();
+      if (vizEntity instanceof NetSimVizWire) {
+        vizEntity.getRoot().prependTo(newParent);
+      } else {
+        vizEntity.getRoot().appendTo(newParent);
+      }
+      vizEntity.onDepthChange(newParent === foreground);
+    }
+  }, this);
+};
+
+/**
+ * Visit method for pullElementsToForeground, not used anywhere else.
+ * Notes that the current entity is should be foreground when we're all done,
+ * finds the current entity's unvisited connections,
+ * pushes those connections onto the stack.
+ * @param {NetSimVizEntity} entityBeingVisited
+ * @param {Array.<NetSimVizEntity>} stack
+ * @private
+ */
+NetSimVisualization.prototype.visitEntityToSetForeground_ = function (
+    entityBeingVisited, stack) {
+  var typedEntity; // Used to mitigate JSDoc complaints.
+
+  // Visit the entity
+  entityBeingVisited.speculativeIsForeground = true;
+
+  // Push new entities to explore based on node type and connections
+  if (entityBeingVisited instanceof NetSimVizNode) {
+    typedEntity = /** @type {NetSimVizNode} */ (entityBeingVisited);
```

Comment:
This is effectively a JSDoc cast.  RubyMine was complaining; but this is Javascript, and it's probably better to jump through fewer hoops even if my editor complains.  I'll change this back.